src/client.rs

```diff
@@ -25,29 +27,127 @@ impl Connectable for &str {
     }
 }
 
-pub enum Connection {
+pub struct Connection {
+    pub(crate) connection: ConnectionInner,
+    tainted: bool,
+    read_timeout: Option<Duration>,
+    write_timeout: Option<Duration>,
```

Comment:
I wonder if we should have rename these two timeouts differently, because they are currently implemented as application-level timeout, not standard stream timeout i.e https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_read_timeout

Maybe change this to `memcache_read_timeout/memcache_write_timeout` ? Or alternatively, just `memcache_timeout` and use the same timeout across all operations. Do we ever expect these two timeouts to be different ?
And also it's probably reasonable to put some docstring here to explain the timeouts and which operation they affect.

Reply:
One may also expect that `read_timeout/write_timeout` should actually propagate to the tcp/unix stream as well. Have we tried that ?

Reply:
> One may also expect that read_timeout/write_timeout should actually propagate to the tcp/unix stream as well. Have we tried that ?

Yes, its a no-op on tokio's tcp/unix stream since they're running in non blocking mode. 

> Or alternatively, just memcache_timeout and use the same timeout across all operations. Do we ever expect these two timeouts to be different ?

Yes, we do.

> I wonder if we should have rename these two timeouts differently, because they are currently implemented as application-level timeout, not standard stream timeout i.e https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_read_timeout
And also it's probably reasonable to put some docstring here to explain the timeouts and which operation they affect.

Sounds good.